test/custom-task-ctrls/wait-task/pkg/reconciler/reconciler_test.go

```diff
@@ -260,3 +232,152 @@ status:
 	}
 }
+
+func TestReconcile_Retries(t *testing.T) {
+	for _, tc := range []struct {
+		name                string
+		duration            string
+		timeout             string
+		retries             int
+		params              string
+		wantStatus          string
+		isParentPRCancelled bool
+	}{{
+		name:     "retry when timeout",
+		duration: "2s",
+		timeout:  "1s",
+		retries:  1,
+		wantStatus: fmt.Sprintf(`
+status:
+  conditions:
+  - reason: %s
+    status: %q
+    type: %s
+  observedGeneration: 0
+  retriesStatus:
+  - conditions:
+    - reason: %s
+      status: %q
+      type: %s
+`, "TimedOut", corev1.ConditionFalse, apis.ConditionSucceeded, "TimedOut", corev1.ConditionFalse, apis.ConditionSucceeded),
+		isParentPRCancelled: false,
+	}, {
+		name:     "don't retry if retries unspecified",
+		duration: "2s",
+		timeout:  "1s",
+		wantStatus: fmt.Sprintf(`
+status:
+  conditions:
+  - reason: %s
+    status: %q
+    type: %s
+  observedGeneration: 0
+`, "TimedOut", corev1.ConditionFalse, apis.ConditionSucceeded),
+		isParentPRCancelled: false,
+	}, {
+		name:     "don't retry when canceled",
+		duration: "2s",
+		wantStatus: fmt.Sprintf(`
+status:
+  conditions:
+  - reason: %s
+    status: %q
+    type: %s
+  observedGeneration: 0
+`, "Cancelled", corev1.ConditionFalse, apis.ConditionSucceeded),
+		isParentPRCancelled: true,
+	}} {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx := context.Background()
+			rec := &Reconciler{
+				Clock: testClock,
+			}
+
+			runName := helpers.ObjectNameForTest(t)
+			runYAML := fmt.Sprintf(`
+metadata:
+  name: %s
+spec:
+  retries: %d
+  timeout: %s
+  ref:
+    apiVersion: %s
+    kind: %s
+  params:
+  - name: duration
+    value: %s
+`, runName, tc.retries, tc.timeout, apiVersion, kind, tc.duration)
+			r := parse.MustParseRun(t, runYAML)
+			if tc.isParentPRCancelled {
+				r.Spec.Status = v1alpha1.RunSpecStatusCancelled
+				r.Spec.StatusMessage = v1alpha1.RunCancelledByPipelineMsg
+			}
+
+			fakeReconcile(t, ctx, rec, r)
+
+			// Compose expected Run
+			wantRunYAML := fmt.Sprintf(`
+metadata:
+  name: %s
+spec:
+  retries: %d
+  timeout: %s
+  ref:
+    apiVersion: %s
+    kind: %s
+  params:
+  - name: duration
+    value: %s
+`, runName, tc.retries, tc.timeout, apiVersion, kind, tc.duration)
+			wantRunYAML = wantRunYAML + tc.wantStatus
+			wantRun := parse.MustParseRun(t, wantRunYAML)
+			if tc.isParentPRCancelled {
+				wantRun.Spec.Status = v1alpha1.RunSpecStatusCancelled
+				wantRun.Spec.StatusMessage = v1alpha1.RunCancelledByPipelineMsg
+			}
+
+			if d := cmp.Diff(wantRun, r,
+				filterTypeMeta,
+				filterObjectMeta,
+				filterCondition,
+				filterRunStatus,
+			); d != "" {
+				t.Errorf("-got +want: %v", d)
+			}
+		})
+	}
+}
+
+func fakeReconcile(t *testing.T, ctx context.Context, rec *Reconciler, r *v1alpha1.Run) {
+	t.Helper()
+
+	// Start reconciling the Run.
+	// This will not return until the second Reconcile is done.
+	err := rec.ReconcileKind(ctx, r)
+
+	t.Logf("%#v", r)
```

Comment:
Is this for testing purposes during development? If so, shall we remove it?

Reply:
SG! updated!